### PR TITLE
auth: Added GoogleOIDCServiceAccount auth source support.

### PR DESCRIPTION
**Pull request by Bplotka**
Signed-off-by: Bartek Plotka <bwplotka@gmail.com>

### CODE REVIEW

File: pkg/tokenauth/sources/oidc/oidc.go [hunk 1]
```diff
@@ -43,11 +43,33 @@ func NewWithCache(name string, cache login.Cache, callbackSrv *login.CallbackSer
...
+		googleServiceAccountJSON,
+		config.Provider,
+		gsa.OIDCConfig{
+			ClientID:     config.ClientID,
+			ClientSecret: config.ClientSecret,
+			Scopes:       config.Scopes,
+		},
+	)
+	if err != nil {
+		return nil, err
```
Comment (stefan-improbable) [inline pkg/tokenauth/sources/oidc/oidc.go]: Can you wrap this error and add more info here?

File: pkg/tokenauth/sources/oidc/oidc.go [hunk 2]
```diff
@@ -43,11 +43,33 @@ func NewWithCache(name string, cache login.Cache, callbackSrv *login.CallbackSer
...
 		name:        name,
-		cache:       cache,
 		tokenSource: tokenSource,
 	}, clearIDTokenFunc, nil
 }
 
+// NewGoogleFromServiceAccount constructs tokenauth.Source that is able to return valid OIDC token from given Google Service Account.
+func NewGoogleFromServiceAccount(name string, config login.OIDCConfig, googleServiceAccountJSON []byte) (tokenauth.Source, error) {
+	tokenSource, _, err := gsa.NewOIDCTokenSource(
+		context.Background(),
```
Comment (stefan-improbable) [inline pkg/tokenauth/sources/oidc/oidc.go]: I tihnk context.Background doesn't have any timeout. Cany ou check whether this is the case and fix it if it its?

File: pkg/winch/auth.go [hunk 3]
```diff
@@ -69,7 +71,22 @@ func (f *AuthFactory) Get(configSource *pb.AuthSource) (tokenauth.Source, error)
 		)
 		// Register handler for clearing ID token.
 		f.mux.HandleFunc(fmt.Sprintf("/winch/cleartoken/%s", configSource.Name), oidcClearTokenHandler(clearIDTokenFunc))
+	case *pb.AuthSource_ServiceAccountOidc:
+		serviceAccountJson, err := ioutil.ReadFile(s.ServiceAccountOidc.ServiceAccountJsonPath)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to open Google SA from %s", s.ServiceAccountOidc.ServiceAccountJsonPath)
```
Comment (stefan-improbable) [inline pkg/winch/auth.go]: failed to read google SA info from

File: proto/winch/config/auth.proto [hunk 4]
```diff
@@ -19,6 +19,7 @@ message AuthSource {
         KubernetesAccess kube = 3;
         OIDCAccess oidc = 4;
         TokenAccess token = 5;
+        GoogleServiceAccountOIDCAccess service_account_oidc = 6;
```
Comment (stefan-improbable) [inline proto/winch/config/auth.proto]: Can you add more comments for on how do these things work? For example I don't know what is the difference between OIDCAccess and GoogleOIDCAccess.

File: pkg/tokenauth/sources/oidc/oidc.go (hunk 1 above)
Comment (Bplotka) [reply pkg/tokenauth/sources/oidc/oidc.go]: yup

File: pkg/tokenauth/sources/oidc/oidc.go (hunk 2 above)
Comment (Bplotka) [reply pkg/tokenauth/sources/oidc/oidc.go]: good point will fix everywhere in this package

File: pkg/winch/auth.go (hunk 3 above)
Comment (Bplotka) [reply pkg/winch/auth.go]: :+1:

File: proto/winch/config/auth.proto (hunk 4 above)
Comment (Bplotka) [reply proto/winch/config/auth.proto]: It is described in the proto.. under `GoogleServiceAccountOIDCAccess` and `OIDCAccess`

File: pkg/winch/auth.go [hunk 5]
```diff
@@ -42,9 +47,13 @@ func (f *AuthFactory) Get(configSource *pb.AuthSource) (tokenauth.Source, error)
 	var source tokenauth.Source
 	var err error
 
+	// Get is always use on startup only, so we can set reasonable timeout here.
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
```
Comment (stefan-improbable) [inline pkg/winch/auth.go]: This should probably be a flag, might be erroring out on slow networks (?)

Comment (Bplotka) [reply pkg/winch/auth.go]: yea

File: pkg/k8s/flags.go [hunk 6]
```diff
@@ -75,9 +78,11 @@ func NewFromFlags() (*APIClient, error) {
 
 	var source tokenauth.Source
 
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
```
Comment (stefan-improbable) [inline pkg/k8s/flags.go]: maek this flag too lpz

**Pull request merged**